src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,228 @@
+        {
+            if (!disposing)
+            {
+                return;
+            }
+
+            lock (Lock)
+            {
+                if (CommonDisposeUnsynchronized())
+                {
+                    return;
+                }
+
+                foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in _limiters)
+                {
+                    limiter.Value.Value.Dispose();
+                }
+                _limiters.Clear();
+            }
+        }
+
+        protected override async ValueTask DisposeAsyncCore()
+        {
+            Dictionary<TKey, Lazy<RateLimiter>> limiters;
+            lock (Lock)
+            {
+                if (CommonDisposeUnsynchronized())
+                {
+                    return;
+                }
+
+                limiters = _limiters;
+                _limiters = new();
+            }
+
+            foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in limiters)
+            {
+                await limiter.Value.Value.DisposeAsync().ConfigureAwait(false);
+            }
+
+            limiters.Clear();
+
+            return;
+        }
+
+        // This handles the common state changes that Dipose and DisposeAsync need to do, the individual limiters still need to be Disposed after this call
+        private bool CommonDisposeUnsynchronized()
+        {
+            if (_disposed)
+            {
+                return true;
+            }
+            _disposed = true;
+
+            return false;
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(PartitionedRateLimiter));
+            }
+        }
+
+        private static void Replenish(object? state)
+        {
+            DefaultPartitionedRateLimiter<TResource, TKey> limiter = (state as DefaultPartitionedRateLimiter<TResource, TKey>)!;
+            Debug.Assert(limiter is not null);
+
+            if (Interlocked.Exchange(ref limiter._executingTimer, 1) != 0)
+            {
+                // Timer already running, do nothing
+                // This might be an indication that there is a slow ReplenishingRateLimiter.TryReplenish implementation
+                // Or many limiters that need updating and a short timer period
+                // Or threadpool exhaustion unrelated to this type directly
+                // We might want to write an event for this in the future so it's visible to diagnostic tools
+                return;
+            }
+
+            lock (limiter.Lock)
+            {
+                if (limiter._disposed)
+                {
+                    limiter._timer!.Dispose();
+                    limiter._cachedLimiters.Clear();
+                    return;
+                }
+
+                // If the cache has been invalidated we need to recreate it
+                if (limiter._cacheInvalid)
+                {
```

Comment:
Don't we need to clear `limiter._cachedLimiters` here?

Reply:
We don't have removal of limiters yet, that'll be in the next PR (or the one after that)

Reply:
Sure. But this will end up adding duplicates, right?